In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
import pandas as pd
path= '/content/drive/MyDrive/Colab aprendizaje supervisado/Taller extra/insurance.csv'
df= pd.read_csv(path)

In [5]:
df.head()

,age,sex,bmi,children,smoker,region,charges
0,19,female,27.900,0,yes,southwest,16884.92400
1,18,male,33.770,1,no,southeast,1725.55230
2,28,male,33.000,3,no,southeast,4449.46200
3,33,male,22.705,0,no,northwest,21984.47061
4,32,male,28.880,0,no,northwest,3866.85520


In [6]:
df.describe()

,age,bmi,children,charges
count,1338.000000,1338.000000,1338.000000,1338.000000
mean,39.207025,30.663397,1.094918,13270.422265
std,14.049960,6.098187,1.205493,12110.011237
min,18.000000,15.960000,0.000000,1121.873900
25%,27.000000,26.296250,0.000000,4740.287150
50%,39.000000,30.400000,1.000000,9382.033000
75%,51.000000,34.693750,2.000000,16639.912515
max,64.000000,53.130000,5.000000,63770.428010


In [7]:
df.region.value_counts()

,count
region,
southeast,364
southwest,325
northwest,325
northeast,324


In [8]:
df['smoker'] = df['smoker'].replace({'yes': 1, 'no': 0})
df['sex'] = df['sex'].replace({'male': 1, 'female': 0})
df.head()

/tmp/ipython-input-3818855470.py:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['smoker'] = df['smoker'].replace({'yes': 1, 'no': 0})
/tmp/ipython-input-3818855470.py:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['sex'] = df['sex'].replace({'male': 1, 'female': 0})


,age,sex,bmi,children,smoker,region,charges
0,19,0,27.900,0,1,southwest,16884.92400
1,18,1,33.770,1,0,southeast,1725.55230
2,28,1,33.000,3,0,southeast,4449.46200
3,33,1,22.705,0,0,northwest,21984.47061
4,32,1,28.880,0,0,northwest,3866.85520


In [9]:
df['SW'] = np.where(df['region'] == 'southwest', 1, 0)
df['SE'] = np.where(df['region'] == 'southeast', 1, 0)
df['NW'] = np.where(df['region'] == 'northwest', 1, 0)
df['NE'] = np.where(df['region'] == 'northeast', 1, 0)
display(df.head())

,age,sex,bmi,children,smoker,region,charges,SW,SE,NW,NE
0,19,0,27.900,0,1,southwest,16884.92400,1,0,0,0
1,18,1,33.770,1,0,southeast,1725.55230,0,1,0,0
2,28,1,33.000,3,0,southeast,4449.46200,0,1,0,0
3,33,1,22.705,0,0,northwest,21984.47061,0,0,1,0
4,32,1,28.880,0,0,northwest,3866.85520,0,0,1,0


In [10]:
df = df.drop('region', axis=1)
display(df.head())

,age,sex,bmi,children,smoker,charges,SW,SE,NW,NE
0,19,0,27.900,0,1,16884.92400,1,0,0,0
1,18,1,33.770,1,0,1725.55230,0,1,0,0
2,28,1,33.000,3,0,4449.46200,0,1,0,0
3,33,1,22.705,0,0,21984.47061,0,0,1,0
4,32,1,28.880,0,0,3866.85520,0,0,1,0


In [11]:
df.isnull().sum()

,0
age,0
sex,0
bmi,0
children,0
smoker,0
charges,0
SW,0
SE,0
NW,0
NE,0


# Partición entre entrenamiento y prueba

In [12]:
train= df.sample(frac=0.7, random_state=42)
test= df.drop(train.index)

In [13]:
train.head()

,age,sex,bmi,children,smoker,charges,SW,SE,NW,NE
764,45,0,25.175,2,0,9095.06825,0,0,0,1
887,36,0,30.020,0,0,5272.17580,0,0,1,0
890,64,0,26.885,0,1,29330.98315,0,0,1,0
1293,46,1,25.745,3,0,9301.89355,0,0,1,0
259,19,1,31.920,0,1,33750.29180,0,0,1,0


In [14]:
from sklearn.model_selection import train_test_split

X = df.drop('charges', axis=1)
y = df['charges']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [15]:
from sklearn.linear_model import LinearRegression
import numpy as np

# Para usar las variables de la regresión, estas deben ser tipo array
X = np.array(X_train)
Y = np.array(y_train).reshape(-1, 1)

# Modelo de regresión
modelo = LinearRegression()
modelo.fit(X, Y)

# Ver coeficientes
print("Intercepto (a):", modelo.intercept_)
print("Pendiente (b):", modelo.coef_)

Intercepto (a): [-12970.02049227]
Pendiente (b): [[  261.29692414   104.81182299   348.90691516   424.11912829
  23628.36722236  -330.26631867  -374.91224933   109.12197876
    596.05658924]]


In [16]:
df_train = pd.concat([y_train, X_train], axis=1)
df_train.head()

,charges,age,sex,bmi,children,smoker,SW,SE,NW,NE
332,13429.03540,61,0,31.160,0,0,0,0,1,0
355,24603.04837,46,1,27.600,0,0,1,0,0,0
138,27322.73386,54,0,31.900,3,0,0,1,0,0
381,42303.69215,55,1,30.685,0,1,0,0,0,1
292,42112.23560,25,1,45.540,2,1,0,1,0,0


In [17]:
corr_matrix = df_train.corr()
corr_matrix

,charges,age,sex,bmi,children,smoker,SW,SE,NW,NE
charges,1.000000,0.288804,0.040390,0.194264,0.079993,0.778989,-0.017719,0.047119,-0.038769,0.008044
age,0.288804,1.000000,-0.013154,0.106366,0.061936,-0.046504,0.014612,-0.019003,0.014390,-0.009283
sex,0.040390,-0.013154,1.000000,0.002457,0.018857,0.047368,-0.019327,-0.039575,0.042582,0.016671
bmi,0.194264,0.106366,0.002457,1.000000,-0.004898,-0.003923,0.029865,0.250506,-0.125689,-0.156995
children,0.079993,0.061936,0.018857,-0.004898,1.000000,0.023334,-0.019050,-0.021264,0.026750,0.013583
smoker,0.778989,-0.046504,0.047368,-0.003923,0.023334,1.000000,-0.013916,0.036953,-0.034933,0.010780
SW,-0.017719,0.014612,-0.019327,0.029865,-0.019050,-0.013916,1.000000,-0.331108,-0.317348,-0.328355
SE,0.047119,-0.019003,-0.039575,0.250506,-0.021264,0.036953,-0.331108,1.000000,-0.337948,-0.349670
NW,-0.038769,0.014390,0.042582,-0.125689,0.026750,-0.034933,-0.317348,-0.337948,1.000000,-0.335138
NE,0.008044,-0.009283,0.016671,-0.156995,0.013583,0.010780,-0.328355,-0.349670,-0.335138,1.000000


In [18]:
import statsmodels.api as sm

# Intercepto
X_const = sm.add_constant(X_train)

# Modelo de regresión
modelo_sm = sm.OLS(y_train, X_const).fit()

# Imprimir el resumen del modelo (incluye la ecuación y ANOVA)
print(modelo_sm.summary())

# Obtener y mostrar los residuales
residuales = modelo_sm.resid
print("\nResiduales:")
print(residuales)

                            OLS Regression Results                            
Dep. Variable:                charges   R-squared:                       0.742
Model:                            OLS   Adj. R-squared:                  0.740
Method:                 Least Squares   F-statistic:                     333.9
Date:                Thu, 14 Aug 2025   Prob (F-statistic):          6.51e-267
Time:                        19:33:34   Log-Likelihood:                -9492.8
No. Observations:                 936   AIC:                         1.900e+04
Df Residuals:                     927   BIC:                         1.905e+04
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -1.038e+04    965.337    -10.749      0.0

# Otras métricas para evaluar el desempeño del modelo

In [19]:
current_aic = modelo_sm.aic
current_bic = modelo_sm.bic
print("aic:", current_aic)
print("bic:", current_bic)

aic: 19003.690833573528
bic: 19047.265372861828


# Forward selection

In [20]:
def forward_selection(X_train, y_train):

    selected_features = []
    remaining_features = list(X_train.columns)
    best_aic = float('inf')

    while remaining_features:
        best_new_feature = None
        improved = False

        for feature in remaining_features:
            current_features = selected_features + [feature]
            X_temp = sm.add_constant(X_train[current_features])
            model = sm.OLS(y_train, X_temp).fit()
            current_aic = model.aic

            if current_aic < best_aic:
                best_aic = current_aic
                best_new_feature = feature
                improved = True

        if improved:
            selected_features.append(best_new_feature)
            remaining_features.remove(best_new_feature)
            print(f"Añadiendo {best_new_feature}. Current AIC: {best_aic:.2f}")
        else:
            break

    return selected_features

selected_features_forward = forward_selection(X_train, y_train)
print("\nVariables seleccionadas usando forward selection:", selected_features_forward)

Añadiendo smoker. Current AIC: 19385.47
Añadiendo age. Current AIC: 19093.83
Añadiendo bmi. Current AIC: 19004.22
Añadiendo children. Current AIC: 18999.47
Añadiendo NE. Current AIC: 18998.62

Variables seleccionadas usando forward selection: ['smoker', 'age', 'bmi', 'children', 'NE']


# Backward elimination

In [21]:
def backward_elimination(X_train, y_train):

    selected_features = list(X_train.columns)
    X_temp_initial = sm.add_constant(X_train[selected_features])
    model_initial = sm.OLS(y_train, X_temp_initial).fit()
    best_aic = model_initial.aic
    print(f"AIC inicial with all features: {best_aic:.2f}")

    while selected_features:
        feature_to_remove = None
        improved = False
        worst_aic = best_aic # Inicializar con el mejor AIC

        for feature in selected_features:
            temp_features = [f for f in selected_features if f != feature]
            if not temp_features:
                 X_temp = pd.DataFrame({'const': np.ones(len(X_train))})
            else:
                X_temp = sm.add_constant(X_train[temp_features])

            model = sm.OLS(y_train, X_temp).fit()
            current_aic = model.aic

            # En backward elimination, se eliminan las variables que no están, cuando se obtuvo el mejor AIC
            if current_aic < best_aic:
                best_aic = current_aic
                feature_to_remove = feature
                improved = True

        if improved:
            selected_features.remove(feature_to_remove)
            print(f"Variables removidas {feature_to_remove}. Nuevo mejor AIC: {best_aic:.2f}")
        else:
            break

    return selected_features

selected_features_backward = backward_elimination(X_train, y_train)
print("\nVariables seleccionadas usando backward elimination:", selected_features_backward)

AIC inicial with all features: 19003.69
Variables removidas sex. Nuevo mejor AIC: 19001.76

Variables seleccionadas usando backward elimination: ['age', 'bmi', 'children', 'smoker', 'SW', 'SE', 'NW', 'NE']


# Stepwise selection

In [22]:
def stepwise_selection(X_train, y_train):

    selected_features = []
    remaining_features = list(X_train.columns)
    best_aic = float('inf')
    no_improvement_count = 0

    while no_improvement_count < 2: # Continue as long as there is potential for improvement (add or remove)
        added_feature = None
        removed_feature = None

        # Step 1: Attempt to add a feature (Forward step)
        best_new_feature = None
        improved_add = False
        current_best_aic_add = best_aic

        for feature in remaining_features:
            current_features = selected_features + [feature]
            X_temp = sm.add_constant(X_train[current_features])
            # Ensure indices are aligned
            X_temp = X_temp.align(y_train, join='inner', axis=0)[0]
            model = sm.OLS(y_train, X_temp).fit()
            current_aic = model.aic

            if current_aic < current_best_aic_add:
                current_best_aic_add = current_aic
                best_new_feature = feature
                improved_add = True

        if improved_add and current_best_aic_add < best_aic:
            best_aic = current_best_aic_add
            selected_features.append(best_new_feature)
            remaining_features.remove(best_new_feature)
            added_feature = best_new_feature
            print(f"Added {best_new_feature}. Current AIC: {best_aic:.2f}")
            no_improvement_count = 0 # Reset counter if improvement made

        # Step 2: Attempt to remove a feature (Backward step)
        feature_to_remove = None
        improved_remove = False
        current_best_aic_remove = best_aic

        if selected_features: # Only attempt removal if there are selected features
             for feature in selected_features:
                temp_features = [f for f in selected_features if f != feature]
                if not temp_features: # Handle the case where removing the last feature
                    X_temp = pd.DataFrame({'const': np.ones(len(X_train))}, index=X_train.index)
                else:
                    X_temp = sm.add_constant(X_train[temp_features])

                # Ensure indices are aligned
                X_temp = X_temp.align(y_train, join='inner', axis=0)[0]
                model = sm.OLS(y_train, X_temp).fit()
                current_aic = model.aic

                if current_aic < current_best_aic_remove:
                    current_best_aic_remove = current_aic
                    feature_to_remove = feature
                    improved_remove = True

        if improved_remove and current_best_aic_remove < best_aic:
            best_aic = current_best_aic_remove
            remaining_features.append(feature_to_remove)
            selected_features.remove(feature_to_remove)
            removed_feature = feature_to_remove
            print(f"Removed {feature_to_remove}. Current AIC: {best_aic:.2f}")
            no_improvement_count = 0 # Reset counter if improvement made

        # Check for improvement in this iteration
        if added_feature is None and removed_feature is None:
            no_improvement_count += 1
        else:
            no_improvement_count = 0


    return selected_features

selected_features_stepwise = stepwise_selection(X_train, y_train)
print("\nSelected features using stepwise selection:", selected_features_stepwise)

Added smoker. Current AIC: 19385.47
Added age. Current AIC: 19093.83
Added bmi. Current AIC: 19004.22
Added children. Current AIC: 18999.47
Added NE. Current AIC: 18998.62

Selected features using stepwise selection: ['smoker', 'age', 'bmi', 'children', 'NE']


# Sequential Future Selection

In [23]:
from sklearn.linear_model import LinearRegression
from mlxtend.feature_selection import SequentialFeatureSelector as SFS
from sklearn.datasets import make_regression
import statsmodels.api as sm

# Modelo base
lr = LinearRegression()

# Forward selection
sfs = SFS(lr,
          k_features='best',   # Puede ser número fijo o 'best'
          forward=True,        # True = forward, False = backward
          floating=False,      # True para stepwise mixto
          scoring='r2',
          cv=5)

sfs = sfs.fit(X_train, y_train)

print("Variables seleccionadas usando SequentialFeatureSelector (Forward):", sfs.k_feature_names_)

# Crear y ajustar el modelo OLS con las variables seleccionadas
X_train_selected = sm.add_constant(X_train[list(sfs.k_feature_names_)])
model_sfs = sm.OLS(y_train, X_train_selected).fit()

# Resumen del modelo OLS
print("\nResumen del modelo OLS con variables seleccionadas por SequentialFeatureSelector:")
print(model_sfs.summary())

Variables seleccionadas usando SequentialFeatureSelector (Forward): ('age', 'bmi', 'children', 'smoker', 'NE')

Resumen del modelo OLS con variables seleccionadas por SequentialFeatureSelector:
                            OLS Regression Results                            
Dep. Variable:                charges   R-squared:                       0.742
Model:                            OLS   Adj. R-squared:                  0.741
Method:                 Least Squares   F-statistic:                     535.3
Date:                Thu, 14 Aug 2025   Prob (F-statistic):          9.51e-271
Time:                        19:33:35   Log-Likelihood:                -9493.3
No. Observations:                 936   AIC:                         1.900e+04
Df Residuals:                     930   BIC:                         1.903e+04
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                

# Análisis métricas de desempeño

In [24]:
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
import numpy as np
import statsmodels.api as sm

# Predicciones en el set de prueba con el modelo entrenado con SFS features
X_test_selected = sm.add_constant(X_test[list(sfs.k_feature_names_)])
y_pred = model_sfs.predict(X_test_selected)

# RMSE
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
print(f"RMSE: {rmse:.2f}")

# MAE
mae = mean_absolute_error(y_test, y_pred)
print(f"MAE: {mae:.2f}")

# Calculate MAPE
# epsilon para evitar la division por cero
epsilon = 1e-10
mape = np.mean(np.abs((y_test - y_pred) / (y_test + epsilon))) * 100
print(f"MAPE: {mape:.2f}%")

# R2
r2 = r2_score(y_test, y_pred)
print(f"R2: {r2:.2f}")

# R2 ajustado
n = len(y_test) # numero de observaciones
p = X_test_selected.shape[1] - 1 # numero de predictores (excluyendo la constante)
adj_r2 = 1 - (1 - r2) * (n - 1) / (n - p - 1)
print(f"R2 ajustado: {adj_r2:.2f}")

RMSE: 5819.36
MAE: 4151.31
MAPE: 43.72%
R2: 0.77
R2 ajustado: 0.77


In [25]:
from sklearn.metrics import mean_squared_error, mean_absolute_error
import numpy as np

# Predicciones con el set de prueba usando el modelo de entrenamiento con SFS features
X_test_selected = sm.add_constant(X_test[list(sfs.k_feature_names_)])
y_pred = model_sfs.predict(X_test_selected)

# RMSE
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
print(f"RMSE: {rmse:.2f}")

# MAE
mae = mean_absolute_error(y_test, y_pred)
print(f"MAE: {mae:.2f}")

# MAPE
mape = np.mean(np.abs((y_test - y_pred) / y_test)) * 100
print(f"MAPE: {mape:.2f}%")

RMSE: 5819.36
MAE: 4151.31
MAPE: 43.72%


## **Informe Ejecutivo – Modelo de Regresión OLS para Predicción de Cargos Médicos**

### **1. Resumen General**

El modelo OLS desarrollado, con variables seleccionadas mediante *SequentialFeatureSelector*, busca explicar y predecir el valor de `charges` (cargos médicos).
Los resultados indican que el modelo tiene un **buen poder explicativo (R² = 74,2%)**, lo que significa que explica gran parte de la variabilidad de los cargos, aunque persiste un error considerable en predicciones individuales.

---

### **2. Desempeño del Modelo**

* **R²** = 0.742 → El 74,2% de la variabilidad de los cargos médicos es explicada por las variables incluidas.
* **R² ajustado** = 0.741 → La estabilidad del ajuste indica que las variables elegidas aportan de manera consistente.
* **Estadístico F** = 535.3 (*p* < 0.001) → El conjunto de variables es estadísticamente significativo.

**Métricas de error:**

* **RMSE** = 5.819,36 → En promedio, las predicciones se desvían 5.819 unidades monetarias respecto al valor real.
* **MAE** = 4.151,31 → El error absoluto promedio es de 4.151 unidades.
* **MAPE** = 43,72% → El error porcentual medio es alto: las predicciones individuales pueden desviarse casi la mitad del valor real.

**Implicaciones de estas métricas:**

* El modelo es útil para estimaciones globales y análisis de tendencias, pero no es lo suficientemente preciso para fijar precios individuales de manera exacta.
* El alto MAPE sugiere que el modelo podría beneficiarse de variables adicionales o de técnicas más complejas (p. ej., modelos no lineales o interacciones entre variables).
* El RMSE alto implica que existen casos con errores grandes, posiblemente asociados a valores extremos (outliers).

---

### **3. Variables y su Impacto**

| Variable      | Coef.     | p-valor | Interpretación                                                                                |
| ------------- | --------- | ------- | --------------------------------------------------------------------------------------------- |
| **Constante** | -12949.44 | 0.000   | Punto de partida de `charges` cuando todas las variables son cero.                            |
| **Edad**      | 261.67    | 0.000   | Cada año adicional incrementa `charges` en 261,67 unidades.                                   |
| **IMC (BMI)** | 342.44    | 0.000   | Cada punto de IMC incrementa `charges` en 342,44 unidades.                                    |
| **Hijos**     | 429.79    | 0.010   | Cada hijo eleva `charges` en 429,79 unidades.                                                 |
| **Fumador**   | 23620.96  | 0.000   | Ser fumador incrementa `charges` en más de 23 mil unidades, siendo la variable de mayor peso. |
| **Región**    | 785.37    | 0.093   | Diferencia media por región, significancia marginal.                                          |

---

### **4. Hallazgos Clave**

1. **Fumar** es el factor más determinante, multiplicando considerablemente los costos.
2. **Edad** y **BMI** también influyen de forma clara y directa.
3. El **número de hijos** y la **región** tienen un impacto menor, aunque la región carece de fuerte soporte estadístico.
4. El modelo captura bien las tendencias generales, pero la precisión individual debe mejorarse para aplicaciones críticas.

---

### **5. Recomendaciones Estratégicas**

* Diseñar **políticas de prevención** y programas antitabaco, dado el fuerte impacto del hábito de fumar en los costos médicos.
* Implementar estrategias de **control de peso** para reducir el BMI en la población.
* Revisar y ampliar la base de datos incorporando variables como ingresos, historial clínico o nivel de actividad física.
* Considerar modelos predictivos más avanzados (p. ej., regresión polinómica, *gradient boosting*) para mejorar el MAPE.
* Usar este modelo para **análisis agregados y segmentación**, no para fijar tarifas individuales exactas.

---